### PR TITLE
Remove prefixed `#` symbol from tags

### PR DESCRIPTION
**Pull request by SudharakaP**
Resolve https://github.com/opencollective/opencollective/issues/5529

https://user-images.githubusercontent.com/12435965/174680484-22c12fa1-8d63-475d-abc0-8eb5b0b946b7.mp4

### CODE REVIEW

**Comment by znarf** [comment]
I'll let @Betree review this one when he's back.

File: server/lib/utils.js [hunk 1]
```diff
@@ -492,7 +492,8 @@ export const cleanTags = tags => {
   const cleanTagsList = tags
     .filter(t => Boolean(t)) // Remove null values
     .map(t => t.trim()) // Trim tags
-    .filter(t => t.length > 0); // Remove empty tags
+    .filter(t => t.length > 0) // Remove empty tags
+    .map(t => t.replace(/^#/g, '')); // Remove # prefixes
```
Comment (Betree) [inline server/lib/utils.js:496]: If we do this after the `filter`, then we can end up with empty tags. Suggested:

```suggestion
    .map(t => t.replace(/^#+/g, '')); // Remove # prefixes
    .map(t => t.trim()) // Trim again for empty tags with a # prefix
    .filter(t => t.length > 0) // Remove empty tags
```

Would be great to add some tests in `test/server/lib/utils.test.js` while we're at it.

File: migrations/20220620205322-remove-prefixes-from-tags.js [hunk 2]
```diff
@@ -0,0 +1,13 @@
+'use strict';
+
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize.query(`
+      UPDATE "Collectives" SET tags = array(SELECT regexp_replace(array_to_string(tags,', '), '^#| #| ', '', 'g')) WHERE tags IS NOT NULL AND EXISTS(SELECT * FROM unnest(tags) tag WHERE tag ILIKE '#%');
```
Comment (Betree) [inline migrations/20220620205322-remove-prefixes-from-tags.js:6]: Would be great to indent the query in that kind of place, it's difficult to parse.

The approach of `stringify(tags) |> regex_replace(...) |> to_array(...)` chosen here seems dangerous as it's going to split tags that include a comma (see https://opencollective.com/cutiebipocfest: `queer trans* inter* black, indigenous and people of color`). We need to find a more elegant solution, ideally not stringifying the array.

**Review by Betree — CHANGES_REQUESTED** [review]

File: server/lib/utils.js (hunk 1 above)
Comment (SudharakaP) [reply server/lib/utils.js]: Done, and I've added a test as well. 👍🏽

File: migrations/20220620205322-remove-prefixes-from-tags.js (hunk 2 above)
Comment (SudharakaP) [reply migrations/20220620205322-remove-prefixes-from-tags.js]: I've changed the script to not use `array_to_string`. 👍🏽

File: migrations/20220620205322-remove-prefixes-from-tags.js [hunk 3]
```diff
@@ -0,0 +1,16 @@
+'use strict';
+
+module.exports = {
+  up: async queryInterface => {
+    await queryInterface.sequelize.query(`
+      UPDATE "Collectives" SET tags = array(SELECT regexp_replace(tag, '^#|', '', 'g')
```
Comment (Betree) [inline migrations/20220620205322-remove-prefixes-from-tags.js:6]: I quickly looked at prod data with this query, it looks like it's doing a great job. 168 collectives will be impacted

```sql
SELECT tags, array(SELECT regexp_replace(tag, '^#|', '', 'g') FROM unnest(tags) as tag)
FROM "Collectives"
WHERE tags IS NOT NULL
AND EXISTS(SELECT * FROM unnest(tags) tag WHERE tag ILIKE '#%');
```

<details>

Collective IDs: 54944
878
829
40171
22379
168194
66787
193736
177506
157330
20351
649
30742
161269
68779
196299
167945
94913
333
20108
26594
45937
877
11452
249305
62066
25369
35165
26011
16551
27352
876
249342
76430
12961
12964
36337
56462
694
12963
13171
923
67260
103596
18755
104238
73110
67097
69203
25310
26596
600
207
14802
17135
13168
72563
330
61760
11655
166582
86437
170533
170525
170930
87694
248178
251260
171262
177067
253760
172548
415934
416722
285123
157796
100124
447285
231092
15107
143147
19634
450107
530
452536
13174
21628
460724
10559
45928
26598
16878
466236
17776
466458
20445
61986
734
479201
41772
482849
45789
61830
78871
696
769
828
11332
21410
634
382
334
336
381
575
590
9908
14341
16378
17459
40817
91737
484189
487242
487873
465084
495532
497791
499573
92788
506840
506844
506870
9802
219632
407
514176
510815
510816
465214
32341
52162
48008
34499
123404
53952
34801
441061
425270
470494
49833
456502
502035
140880
127385
51983
285186
453418
429608
164948
519583
25406
523815
523874
445076
526041
143336
533270

</details>

**Review by Betree — APPROVED** [review]
Perfect :+1:

**Pull request merged**